test/cache/preload_test.dart

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:http/http.dart';
+import 'package:path/path.dart' as p;
+import 'package:pub/src/exit_codes.dart';
+import 'package:test/test.dart';
+
+import '../descriptor.dart' as d;
+import '../descriptor.dart';
+import '../test_pub.dart';
+
+void main() {
+  test('adds correct entries to cache and stores the content-hash', () async {
+    final server = await servePackages();
+    server.serve('foo', '1.0.0');
+    server.serve('foo', '2.0.0');
+
+    await appDir({'foo': '^2.0.0'}).create();
+    // Do a `pub get` here to create a lock file in order to validate we later can
+    // `pub get --offline` with packages installed by `preload`.
+    await pubGet();
+
+    await runPub(args: ['cache', 'clean', '-f']);
+
+    final archivePath = p.join(sandbox, 'archive');
+    final archivePath2 = p.join(sandbox, 'archive2');
+
+    File(archivePath).writeAsBytesSync(await readBytes(
+        Uri.parse(server.url).resolve('packages/foo/versions/1.0.0.tar.gz')));
+    File(archivePath2).writeAsBytesSync(await readBytes(
+        Uri.parse(server.url).resolve('packages/foo/versions/2.0.0.tar.gz')));
```

Comment:
nit: but are you sure you don't want a trailing comma here :rofl: (do what you like, I can't really care :see_no_evil: )
```suggestion
    File(archivePath).writeAsBytesSync(
      await readBytes(Uri.parse(server.url).resolve('packages/foo/versions/1.0.0.tar.gz')),
    );
    File(archivePath2).writeAsBytesSync(
      await readBytes(Uri.parse(server.url).resolve('packages/foo/versions/2.0.0.tar.gz')),
    );
```